libbeat/publisher/queue/wrapper.go

```diff
@@ -0,0 +1,93 @@
+package queue
+
+import (
+	"sync"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/publisher"
+)
+
+// Wraps Queue interface to add QoS functionality on top of it.
+type QueueWrapper struct {
+	q   Queue
+	qos Qoser
+}
+
+// TODO: Add config reading capability
+func NewQueueWrapper(q Queue, _ *common.Config) Queue {
+	if q != nil {
+		qw := &QueueWrapper{
+			q:   q,
+			qos: NewWeightedScheduler(),
+		}
+
+		qw.qos.Schedule()
+		return qw
+	}
+	return nil
+}
+
+func (q *QueueWrapper) BufferConfig() BufferConfig {
+	return q.q.BufferConfig()
+}
+
+func (q *QueueWrapper) Producer(cfg ProducerConfig) Producer {
+	return newProducerWrapper(q.q.Producer(cfg), cfg.Weight, q.qos)
+}
+
+func (q *QueueWrapper) Consumer() Consumer {
+	return q.q.Consumer()
+}
+
+func (q *QueueWrapper) Close() error {
+	return q.q.Close()
+}
+
+// Wraps Producer interface to add QoS functionality on top of it.
+type ProducerWrapper struct {
+	QoserClient
+	p Producer
+
+	wg     sync.WaitGroup
+	weight int
+}
+
+func newProducerWrapper(p Producer, weight int, qos Qoser) *ProducerWrapper {
+	if p != nil {
+		return &ProducerWrapper{
+			p:           p,
+			QoserClient: qos.CreateClient(),
+			weight:      weight,
+		}
+	}
+	return nil
+}
+
+func (p *ProducerWrapper) Publish(event publisher.Event) bool {
+	p.waitForPublish(event)
+	return p.p.Publish(event)
+}
+
+func (p *ProducerWrapper) TryPublish(event publisher.Event) bool {
+	p.waitForPublish(event)
+	return p.p.TryPublish(event)
```

Comment:
There is no limit on waitForPublish. `TryPublish` must return `false` immediately if it can not serve the event right now. This is required for packetbeat, so it can continue processing events if publishing is not possible right now.